pkgs/servers/code-server/default.nix

```diff
@@ -0,0 +1,139 @@
+{ stdenv, fetchFromGitHub, python, zip, makeWrapper, fetchurl, pkgconfig, git
+, rsync, makeDesktopItem, runtimeShell, writeScript, runCommand
+, nodejs, yarn, libsecret, xorg, ripgrep, electron }:
+
+with stdenv.lib;
+
+let
+  system = stdenv.hostPlatform.system;
+
+in stdenv.mkDerivation rec {
+  pname = "code-server";
+  version = "3.2.0";
+  commit = "fd36a99a4c78669970ebc4eb05768293b657716f";
+
+  src = fetchFromGitHub {
+    owner = "cdr";
+    repo = "code-server";
+    rev = version;
+    sha256 = "PX2WE4Wr4XUcJFw/FfKZ7PxueMXCOV/qQRGVqPThTqQ=";
+    fetchSubmodules = true;
+  };
+
+  yarnCache = stdenv.mkDerivation {
+    name = "${pname}-${version}-${system}-yarn-cache";
+    inherit src;
+    phases = ["unpackPhase" "buildPhase"];
+    nativeBuildInputs = [ yarn git ];
+    buildPhase = ''
+      export HOME=$PWD
+
+      # apply code-server patches as code-server has patched vscode yarn.lock
+      yarn vscode:patch
+
+      yarn config set yarn-offline-mirror $out
+      find "$PWD" -name "yarn.lock" -printf "%h\n" | \
+        xargs -I {} yarn install --cwd {} \
+          --frozen-lockfile --ignore-scripts --ignore-platform --ignore-engines --no-progress --non-interactive
+    '';
+    outputHashMode = "recursive";
+    outputHashAlgo = "sha256";
+
+    # to get hash values use nix-build -A code-server.yarnPrefetchCache --argstr system <system>
+    outputHash = {
+      x86_64-linux = "197v9660yqhvazalyhc87ciy6q2ii8xxi5bc5y96bck2w0d2y8if";
+      aarch64-linux = "LiIvGuBismWSL2yV2DuKUWDjIzuIQU/VVxtiD4xJ+6Q=";
+    }.${system} or (throw "Unsupported system ${system}");
+  };
+
+  # Extract the Node.js source code which is used to compile packages with
+  # native bindings
+  nodeSources = runCommand "node-sources" {} ''
+    tar --no-same-owner --no-same-permissions -xf ${nodejs.src}
+    mv node-* $out
+  '';
+
+  nativeBuildInputs = [ nodejs yarn python pkgconfig zip makeWrapper git rsync ];
```

Comment:
surely not Python 2 is needed?

Reply:
The [docs on building vscode](https://github.com/Microsoft/vscode/wiki/How-to-Contribute#build-and-run) say

> * [Python](https://www.python.org/downloads/release/python-2715/), at least version 2.7 (version 3 is _**not**_ supported)
>   * **Note:** Python 2.7 will be automatically installed for Windows users through installing `windows-build-tools` npm module (see below)

Curiously the Homebrew formula for code-server declares a dependency on python3.8. I think someone was confused though based on the above, and it probably only works because macOS still ships with python 2 (though it likely won’t in macOS 10.16).

Reply:
Explicitly set python to `python3`

Reply:
I haven't tested this derivation myself, but since the vscode docs explicitly state that python 3 is not supported, don't we need to be using python 2 here?

I also find it confusing that the selection of python 3 is done in the `callPackage` instead of in the arguments to this derivation.

Reply:
Ok, I can change to python2 explicitly and try to rebuild.

Is putting it in attribute set preffered way? The issue is when you want to override for example:

```
code-server.override { python2 = python3; }
```

instead of:

```
code-server.override { python = python3; }
```

The first one feels wierd, so that's my issue with putting in package itself.

Reply:
But it looks like it builds with python3 just fine, why do we need to use python2 for building then?

Reply:
I decided to let it build with python3, as this is only build time dependency, so there should be no issues in runtime, and as long as it builds fine and works fine it's fine for me. Idk what issues vscode has with python3 or if documentation is up to date, but it works fine for us. If there are any issues in the future we can still switch it to python2.

Reply:
@lilyball i implemented suggested changes, except change to `python2`.